### PR TITLE
refactor: Move packages tags to the usbmux for future fixes

### PR DESCRIPTION
**Pull request by umutuzgur**
There some race conditions in protocol upgrades when the phone initiates it before we can change the pipeline. There will be another PR followed by this to address that issue

### CODE REVIEW

File: lib/usbmux/index.js [hunk 1]
```diff
@@ -45,6 +45,20 @@ class Usbmux extends BaseServiceSocket {
...
+
+    this._tag = 0;
+    this._responseCallbacks = {};
+    this._decoder.on('data', this._handleData.bind(this));
+
+  }
+
+  _handleData (data) {
+    const cb = this._responseCallbacks[data.header.tag];
+    if (!_.isFunction(cb)) {
```
Comment (mykola-mokhnach) [inline lib/usbmux/index.js]: I'd rather change it to

```
if (_.isFunction(cb)) {
  cb(data)
}
```

File: lib/usbmux/index.js [hunk 2]
```diff
@@ -101,10 +121,12 @@ class Usbmux extends BaseServiceSocket {
   }
 
   _receivePlistPromise (timeout = 5000) {
-    return new B((resolve, reject) => {
-      this._decoder.once('data', resolve);
+    const tag = this._tag++;
```
Comment (mykola-mokhnach) [inline lib/usbmux/index.js]: can there be a mathematical overflow on this tag?

File: lib/usbmux/index.js [hunk 3]
```diff
@@ -76,13 +93,16 @@ class Usbmux extends BaseServiceSocket {
    * @returns {?Object}
    */
   async readPairRecord (udid, timeout = 5000) {
-    const receivePromise = this._receivePlistPromise(timeout);
+    const { tag, receivePromise } = this._receivePlistPromise(timeout);
```
Comment (mykola-mokhnach) [inline lib/usbmux/index.js]: should we update the docstring on this method?

File: lib/usbmux/index.js [hunk 4]
```diff
@@ -45,6 +45,20 @@ class Usbmux extends BaseServiceSocket {
...
+
+  }
+
+  _handleData (data) {
+    const cb = this._responseCallbacks[data.header.tag];
+    if (!_.isFunction(cb)) {
+      return;
+    }
+    cb(data); // eslint-disable-line promise/prefer-await-to-callbacks
+
```
Comment (mykola-mokhnach) [inline lib/usbmux/index.js]: extra empty line

File: lib/usbmux/index.js [hunk 5]
```diff
@@ -45,6 +45,20 @@ class Usbmux extends BaseServiceSocket {
 
     this._encoder = new UsbmuxEncoder();
     this._encoder.pipe(this._socketClient);
+
+    this._tag = 0;
+    this._responseCallbacks = {};
+    this._decoder.on('data', this._handleData.bind(this));
+
```
Comment (mykola-mokhnach) [inline lib/usbmux/index.js]: extra empty line

File: lib/usbmux/index.js (hunk 2 above)
Comment (umutuzgur) [reply lib/usbmux/index.js]: That would be impressive. I love you how you see stuff :D

File: lib/usbmux/index.js (hunk 3 above)
Comment (umutuzgur) [reply lib/usbmux/index.js]: There is no docstring on it. It is a pretty simple and internal method. I don't think we need it here

**Pull request merged**